feat: Добавлена возможность комментирования администраторами событий,…

IIExpanse opened the pull request:
Администрирование: второй вариант. Выгрузка ожидающих модерацию событий и добавление комментариев к ним.

IIExpanse committed:
feat: Добавлена возможность комментирования администраторами событий, ожидающих модерации.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/controller/EventController.java:
```diff
@@ -101,6 +101,14 @@ public Collection<EventFullDto> getFilteredEventsInternal(
         );
     }
 
+    @GetMapping(path = "/admin/vents/pending")
```
Мне кажется в пути ошибка, должно быть events

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/model/Event.java:
```diff
@@ -64,4 +65,8 @@ public class Event {
+
+    @OneToMany(mappedBy = "event")
+    @ToString.Exclude
+    private Set<Review> reviews;
```
А здесь нужен именно Set? Просто он намного медленнее чем List. 
Еще, если используешь интерфейс Set, то надо смотреть реализацию. Hibernate, вроде, должен вернуть HashSet (хотя не уверен, у него просто есть свои коллекции), тогда надо переопределить Equals и Hashcode в Review, иначе можно рано или поздно поймать утечку памяти OutOfMemory. А если переопределяешь через lombok equals и hashcode, то надо следить за полями, которые туда попадают, чтобы не словить LazyInitialisationExcepton.
Поэтому еще раз вопрос:  "А здесь нужен именно Set?"=))

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/impl/EventServiceImpl.java:
```diff
@@ -198,6 +198,32 @@ public Collection<EventFullDto> getFilteredEventsInternal(
+                        EventState.REVIEWED,
+                        Pageable.ofSize(size))
+                .stream()
+                .skip(from)
```
Можно избавиться от Pageable и передавать PageRequest.of(from, size), тогда репозиторий должен вернуть Page сразу с учетом from и не надо будет запускать stream и скипать.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/EventService.java:
```diff
@@ -43,6 +43,11 @@ Collection<EventFullDto> getFilteredEventsInternal(
+    Collection<EventFullDto> getAllEventsToReview(
+            int from,
+            int size
+    );
```
Вроде, всего 2 аргумента, красивее будет в 1 строчку написать. В других методах этого интерфейса, где 2-3 аргумента все записано в 1 строку.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/impl/EventServiceImpl.java:
```diff
@@ -198,6 +198,32 @@ public Collection<EventFullDto> getFilteredEventsInternal(
+                .skip(from)
+                .collect(Collectors.toSet());
+
+        Collection<Long> eventIds = events.stream().map(Event::getId).collect(Collectors.toSet());
```
Стримы обычно так в дну строчку не пишут. Читаемость теряется. И еще у Collectors.toSet() я бы импортировал статический метод, еще лучше читаемость становится. Лучше делать:
events.stream()
    .map(Event::getId)
    .collect(toSet());

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/impl/EventServiceImpl.java:
```diff
@@ -198,6 +198,32 @@ public Collection<EventFullDto> getFilteredEventsInternal(
+                        Pageable.ofSize(size))
+                .stream()
+                .skip(from)
+                .collect(Collectors.toSet());
```
У Event тоже не переопределен Equals и Hashcode. toSet здесь точно возвращает HashSet.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/impl/EventServiceImpl.java:
```diff
@@ -214,7 +240,7 @@ public EventFullDto updateEvent(UpdateEventRequest request, long userId) {
                     "нельзя отредактировать опубликованное событие.", eventId));
 
-        } else if (state == EventState.CANCELED) {
+        } else if (state == EventState.CANCELED || state == EventState.REVIEWED) {
```
Я бы заимпортил EventState.* Улучшиться читаемость.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/controller/ReviewController.java:
```diff
@@ -0,0 +1,43 @@
+@RequiredArgsConstructor
+@ResponseStatus(HttpStatus.OK)
+public class ReviewController {
+
```
Обычно между декларированием класса и его полями пустых строчек не оставляют.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/controller/ReviewController.java:
```diff
@@ -0,0 +1,43 @@
+
+@RestController
+@RequiredArgsConstructor
+@ResponseStatus(HttpStatus.OK)
```
В контроллере есть @PostMapping, по идее у него ответ должен быть CREATED

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/dto/NewReviewDto.java:
```diff
@@ -0,0 +1,23 @@
+import javax.validation.constraints.NotNull;
+
+@Data
+@Builder
```
Не рекомендуется использовать @Data и @Builder. Вообще, любые сочетания аннотаций, которые генерят одинаковые методы Лучше бы было избегать, как правило все будет работать, но иногда...))

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/dto/ReviewDto.java:
```diff
@@ -0,0 +1,21 @@
+import java.time.LocalDateTime;
+
+@Data
+@Builder
```
Та же самая история. Мой ревьюер присылал  мне статью, где было писано, почему нельзя, но это было в слаке, а в закладки я ее забыл сохранить при переезде((

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/exception/CanOnlyReviewPendingEventsException.java:
```diff
@@ -0,0 +1,8 @@
+public class CanOnlyReviewPendingEventsException extends RuntimeException {
+
+    public CanOnlyReviewPendingEventsException(String message) {
+        super(message);
```
А есть вообще смысл добавлять еще один класс, если происходит просто вызов родителя? Мне кажется можно было бы повесить логгирование или еще что-то, я понимаю, конечно, что еще один класс хлеба не просит, но тут совсем скудный функционал.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/model/Review.java:
```diff
@@ -0,0 +1,32 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "review_id")
+    private Long id;
+    @Column(name = "reviewer_name")
```
Мне кажется, поле имя можно ограничить длинной, условно 100 символов, а не использовать стандартные 255

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/model/Review.java:
```diff
@@ -0,0 +1,32 @@
+    private Long id;
+    @Column(name = "reviewer_name")
+    private String reviewerName;
+    @Column(name = "text")
```
А здесь наоборот хватит ли 255 символов на ревью?

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/model/Review.java:
```diff
@@ -0,0 +1,32 @@
+    @Column(name = "created_on")
+    private LocalDateTime created;
+    @ManyToOne
+    @JoinColumn(name = "reviewed_event", referencedColumnName = "event_id")
```
Обычно name = "event_id", чтобы было понятно, что тут лежит просто референс на другую таблицу.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/model/Review.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "review_id")
```
Не совсем замечание или предложение, но зачем менять название id? С ним же по таблице все понятно, и для @JoinTabele не надо указывать referenceColumnName, а и Join делать удобнее, но это на любителя=))

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+@Slf4j
+@Transactional
+public class ReviewServiceImpl implements ReviewService {
+
```
Здесь тоже лишняя строчка оставлена пустой.

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Override
+    public ReviewDto addNewReview(NewReviewDto reviewDto) {
+        Event event = eventRepository.findById(reviewDto.getEventId()).orElseThrow(() ->
```
orElseThrow лучше вынести на отдельную строку

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Override
+    public Collection<ReviewDto> getEventReviewsByOwner(long userId, long eventId, int from, int size) {
+        Event event = eventRepository.findById(eventId).orElseThrow(() ->
```
Здесь я бы тоже разместил эту конструкцию на одной строке
.orElseThrow(() ->  new EventNotFoundException(

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public Collection<ReviewDto> getEventReviewsInternal(long eventId, int from, int size) {
+        return reviewRepository.findAllByEventIdOrderByCreated(eventId, Pageable.ofSize(size)).stream()
+                .skip(from)
```
Здесь тоже можно избавиться от skip, если передавать PageRequest.of(from, size)

Difirton commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/review/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+        return reviewRepository.findAllByEventIdOrderByCreated(eventId, Pageable.ofSize(size)).stream()
+                .skip(from)
+                .map(mapper::mapToDto)
+                .collect(Collectors.toSet());
```
Про сеты уже написал, тем более, если возвращаешь коллекцию, то лучше List. Все таки все эти проверки в Set на отсутствие дубликатов, они е бесплатны. В любом случае это медленно и требует больше памяти.

Difirton commented on ewm-service/src/main/resources/schema.sql:
```diff
@@ -42,7 +42,7 @@ CREATE TABLE IF NOT EXISTS "events"
     CONSTRAINT event_title_not_blank CHECK (LENGTH("title") > 0),
     CONSTRAINT created_before_publishing CHECK ("created_on" < "published_on"),
-    CONSTRAINT in_event_state_enum CHECK ("state" IN ('PENDING', 'PUBLISHED', 'CANCELED'))
+    CONSTRAINT in_event_state_enum CHECK ("state" IN ('PENDING', 'PUBLISHED', 'CANCELED', 'REVIEWED'))
```
Прикольно, я забыл добавить проверки в БД, возьму себе на заметку=))

Difirton commented on ewm-service/src/main/resources/schema.sql:
```diff
@@ -75,4 +75,17 @@ CREATE TABLE IF NOT EXISTS "events_compilations"
+(
+    "review_id"      BIGINT GENERATED BY DEFAULT AS IDENTITY,
+    "reviewer_name"  VARCHAR   NOT NULL,
+    "text"           VARCHAR   NOT NULL,
```
Мне кажется тут больше подходит тип TEXT, либо более длинный VARCHAR

Difirton reviewed (commented):
Хорошая работа, было интересно посмотреть. Оставил ряд комментариев. 
Желательно обратить внимание на использование Set и использовать его только когда это необходимо. Еще я бы все таки перешел на PageRequest.of(from, size) вместо Pageble, чтобы не гонять лишние данные между БД и микросервисом.

IIExpanse committed:
feat: Добавлены тесты для фичи.
style: Мелкие правки стиля.
fix: Исправлен путь getAllPendingEvents(...) контроллера EventController
